tower-http/src/follow_redirect/mod.rs

```diff
@@ -0,0 +1,325 @@
+    #[pin]
+    future: F,
+    service: S,
+    policy: P,
+    method: Method,
+    uri: Uri,
+    version: Version,
+    headers: HeaderMap<HeaderValue>,
+    body: BodyRepr<B>,
+}
+
+impl<F, S, ReqBody, ResBody, P> Future for ResponseFuture<F, S, ReqBody, P>
+where
+    F: Future<Output = Result<Response<ResBody>, S::Error>>,
+    S: Service<Request<ReqBody>, Future = F> + Clone,
+    ReqBody: Body + Default,
+    P: Policy<ReqBody, S::Error>,
+{
+    type Output = Result<Response<ResBody>, S::Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+        let res = ready!(this.future.as_mut().poll(cx)?);
+
+        match res.status() {
+            StatusCode::MOVED_PERMANENTLY | StatusCode::FOUND => {
+                // User agents MAY change the request method from POST to GET
+                // (RFC 7231 section 6.4.2. and 6.4.3.).
+                if *this.method == Method::POST {
+                    *this.method = Method::GET;
+                    *this.body = BodyRepr::Empty;
+                }
+            }
+            StatusCode::SEE_OTHER => {
+                // A user agent can perform a GET or HEAD request (RFC 7231 section 6.4.4.).
+                if *this.method != Method::HEAD {
+                    *this.method = Method::GET;
+                }
+                *this.body = BodyRepr::Empty;
+            }
+            StatusCode::TEMPORARY_REDIRECT | StatusCode::PERMANENT_REDIRECT => {}
+            _ => return Poll::Ready(Ok(res)),
+        };
+
+        let body = if let Some(body) = this.body.take() {
+            body
+        } else {
+            return Poll::Ready(Ok(res));
+        };
+
+        let location = res
+            .headers()
+            .get(&LOCATION)
+            .and_then(|loc| str::from_utf8(loc.as_bytes()).ok())
+            .and_then(|loc| RiReferenceStr::<UriSpec>::new(loc).ok())
+            .and_then(|loc| {
+                Some(loc.resolve_against(&RiAbsoluteString::try_from(this.uri.to_string()).ok()?))
+            })
+            .and_then(|loc| Uri::try_from(loc.as_str()).ok());
+        let location = if let Some(loc) = location {
+            loc
+        } else {
+            return Poll::Ready(Ok(res));
+        };
+
+        let attempt = Attempt {
+            status: res.status(),
+            location: &location,
+            previous: this.uri,
+        };
+        match this.policy.redirect(&attempt).kind {
+            ActionKind::Follow => {
+                this.body.try_clone_from(&body, &this.policy);
+
+                let mut req = Request::new(body);
+                *req.uri_mut() = location;
+                *req.method_mut() = this.method.clone();
+                *req.version_mut() = *this.version;
+                *req.headers_mut() = this.headers.clone();
+                this.policy.on_request(&mut req);
+                this.future.set(this.service.call(req));
```

Comment:
As `call` is called on the inner service in `FollowRedirect::call` we have to wait for the service to become ready again.

One could do `this.service.clone().oneshot(req)` but that returns a different future type than what `F` is here. Maybe `futures::future::Either` could be used to combine the two types.